### PR TITLE
Stack

### PR DESCRIPTION
**Pull request by DianaVashti**

### CODE REVIEW

File: spec/stack.js [hunk 1]
```diff
@@ -11,12 +11,48 @@ describe('Stack', () => {
     expect(Stack).to.be.a('function')
   })
 
+  context('length()', () => {
+    it('returns the size of the stack', () => {
```
Comment (jrobcodes) [inline spec/stack.js:15]: We should test for possible edge cases as well - does `length` work for empty stacks?

File: spec/stack.js [hunk 2]
```diff
@@ -11,12 +11,48 @@ describe('Stack', () => {
...
+      const myStack = new Stack()
+      myStack.push('foo')
+      myStack.push('bar')
+      expect(myStack.length()).to.eql(2)
+    })
+  })
+
+  context('isEmpty()', () => {
+    it('returns true if the stack is empty', () => {
```
Comment (jrobcodes) [inline spec/stack.js:24]: This test description does not match the expectation under test.  Consider adding both the `true if empty` and `false if not empty` cases.

File: spec/stack.js [hunk 3]
```diff
@@ -11,12 +11,48 @@ describe('Stack', () => {
...
+      myStack.push('foo')
+      myStack.push('bar')
+      expect(myStack.isEmpty()).to.eql(false)
+    })
+  })
+
   context('push()', () => {
     it('pushes an element to the top of the stack.', () => {
       const myStack = new Stack()
+      myStack.push('foo')
```
Comment (jrobcodes) [inline spec/stack.js:35]: You would also want to test that the element pushed was the correct element - this only tests that the stack length changed when pushing.  

Consider implementing `Stack` with a linked list implementation, instead of re-using the native javascript implementation, to ensure familiarity with the behaviors.

File: spec/stack.js [hunk 4]
```diff
@@ -11,12 +11,48 @@ describe('Stack', () => {
...
+      expect(myStack.length()).to.deep.eql(1)
+    })
+  })
+
+  context('pop()', () => {
+    it('pops the top element of the stack', () => {
```
Comment (jrobcodes) [inline spec/stack.js:41]: You should also have a test to ensure that the size of the stack was reduced by 1.

File: spec/stack.js [hunk 5]
```diff
@@ -11,12 +11,48 @@ describe('Stack', () => {
...
+      myStack.push('foo')
+      myStack.push('bar')
+      expect(myStack.pop()).to.eql(['bar'])
+    })
+  })
 
-      expect(() => myStack.push('foo'))
-        .to.alter(() => myStack.length(), { from: 0, to: 1 })
+  context('peek()', () => {
+    it('peeks the top element of the stack', () => {
```
Comment (jrobcodes) [inline spec/stack.js:50]: I might include a test to ensure that the size of the stack *doesn't* change with `peek` (but more because I've seen many students mess this up, than for any other reason 😄 ).

File: src/stack.js [hunk 6]
```diff
@@ -1,5 +1,43 @@
-'use strict'
-
 export default class Stack {
-  // your code here
+  constructor() {
+    this.size = 0
+    this.elements = []
+  }
+
+//stack.length()    // returns the number of elements in the stack.
```
Comment (jrobcodes) [inline src/stack.js:7]: Omit unnecessary comments.

File: src/stack.js [hunk 7]
```diff
@@ -1,5 +1,43 @@
...
+  length() {
+    return this.size
+  }
+
+//stack.isEmpty()   // returns true if the stack is empty or false if not.
```
Comment (jrobcodes) [inline src/stack.js:12]: Omit unnecessary comments.

File: src/stack.js [hunk 8]
```diff
@@ -1,5 +1,43 @@
...
+  isEmpty() {
+    return this.size === 0
+  }
+
+//stack.push("foo") // push an element (the string "foo") to the top of the stack.
+  push(elementToAdd) {
+    this.elementToAdd = elementToAdd
```
Comment (jrobcodes) [inline src/stack.js:19]: I don't understand this implementation?  There's no need to store an instance variable with the last element added.  I'm not sure I see why you are calling `x.split()` if the element type is a `String`, not where `x` comes from.

File: src/stack.js [hunk 9]
```diff
@@ -1,5 +1,43 @@
...
+    this.size += 1
+    if (typeof elementToAdd === String){
+      x.split()
+    }
+    return this.elements = this.elements.concat(elementToAdd)
+  }
+
+
+//stack.pop()       // returns and removes the top element in the stack or null if the stack is empty.
```
Comment (jrobcodes) [inline src/stack.js:28]: Omit unnecessary comments.

File: src/stack.js [hunk 10]
```diff
@@ -1,5 +1,43 @@
...
+  pop() {
+    if (this.size === 0){
+      return null
+    } this.size -= 1
```
Comment (jrobcodes) [inline src/stack.js:32]: Indentation/formatting makes this a little difficult to read.  Consider putting this on a new line, and left aligning it with preceding blocks in its scope (`pop`).

File: src/stack.js [hunk 11]
```diff
@@ -1,5 +1,43 @@
...
+      return this.elements.splice(this.elements.length - 1, 1)
+  }
+
+//stack.peek()      // returns the top element in the stack or null if the stack is empty.
+  peek() {
+    if (this.elements.length === 0){
+      return null
+    } return this.elements[this.elements.length - 1]
```
Comment (jrobcodes) [inline src/stack.js:40]: Indentation/formatting makes this a little difficult to read. Consider putting this on a new line, and left aligning it with preceding blocks in its scope (peek).

**Review by jrobcodes — COMMENTED** [review]
You should add `lib/` to your .gitignore, since the generated code can be generated anywhere the project is installed.
